# Masking Strategy Comparison for PatchTST Self-Supervised Learning

This notebook evaluates the following masking strategies for PatchTST's masked patch reconstruction pretraining on the ETTh1 dataset:
1. Zero masking (baseline): Replace masked patches with zeros
2. Gaussian noise: Add Gaussian noise to masked patches

## Experimental Setup
- Masking ratio: 40%
- Patching configs: `patch_len=12, stride=12` (non-overlapping)
- Pre-training: 1 seed (2021), 10 epochs
- Fine-tuning: 5 seeds (2019-2023)
- Forecast horizons: {96, 192, 336, 720}

## 0. Get Setup

In [ ]:
# Clone the github repo and download dependencies
!git clone https://github.com/chunyagi/PatchTST.git

Cloning into 'PatchTST'...
remote: Enumerating objects: 390, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 390 (delta 9), reused 9 (delta 4), pack-reused 368 (from 1)
Receiving objects: 100% (390/390), 12.94 MiB | 17.29 MiB/s, done.
Resolving deltas: 100% (178/178), done.


In [ ]:
# Install dependencies (uncomment it if you're using google colab)
# !pip install -r PatchTST/PatchTST_supervised/requirements.txt

In [ ]:
# Traverse into the target directory
%cd /content/PatchTST/PatchTST_self_supervised

/content/PatchTST/PatchTST_self_supervised


## 1. Download Data

In [ ]:
# Download data from google drive with gdown
try:
    import gdown
    print(f"[INFO] gdown version: {gdown.__version__}")
except:
    print("[INFO] Couldn't find gdown, installing it...")
    !pip install -q gdown
    import gdown
    print(f"[INFO] gdown version: {gdown.__version__}")

!mkdir -p /data/datasets/public/ETDataset/ETT-small
!gdown 1vOClm_t4RgUf8nqherpTfNnB8rrYq14Q -O /data/datasets/public/ETDataset/ETT-small/

[INFO] gdown version: 5.2.1
Downloading...
From: https://drive.google.com/uc?id=1vOClm_t4RgUf8nqherpTfNnB8rrYq14Q
To: /data/datasets/public/ETDataset/ETT-small/ETTh1.csv
100% 2.59M/2.59M [00:00<00:00, 223MB/s]


## 2. Zero Masking (Baseline)

This is the baseline masking strategy from the original PatchTST paper. It randomly selects patches and replaces their values with zeros.

### 2.1 Pre-training

In [ ]:
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model_id 1 --seed 2019
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model_id 2 --seed 2020
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model_id 3 --seed 2021
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model_id 4 --seed 2022
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model_id 5 --seed 2023

args: Namespace(dset_pretrain='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, mask_ratio=0.4, use_gaussian_noise=0, noise_std=0.1, n_epochs_pretrain=10, lr=0.0001, pretrained_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 17260
suggested_lr 0.0009111627561154895
number of patches: 42
number of model params 17260
          epoch     train_loss     valid_loss           time
Better model found at epoch 0 with valid_loss value: 0.9803512819793537.
              0       1.058250       0.980351          00:01
Better model found at epoch 1 with valid_loss value: 0.9311756620287253.
              1       0.984903       0.931176          00:01
Better model found at epoch 2 with valid_loss value: 0.9152354269524237.
              2       0.955201       0.915235       

### 2.2 Fine-tuning

#### 2.2.1 Horizon 96

In [ ]:
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth successfully transferred!

suggested_lr 0.0027825594022071257
end-to-end finetuning
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pr

In [ ]:
# Evaluate the results
import numpy as np
scores_mse = np.array([0.3738432, 0.37373018, 0.37766117, 0.37854013, 0.3728617])
scores_mae = np.array([0.4026461, 0.40272796, 0.40765455, 0.4067808, 0.40395078])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.375 ± 0.002 | MAE: 0.405 ± 0.002


#### 2.2.2 Horizon 192

In [ ]:
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=192, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 146272
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth successfully transferred!

suggested_lr 0.000756463327554629
end-to-end finetuning
number of patches: 42
number of model params 146272
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-

In [ ]:
# Evaluate the results
scores_mse = np.array([0.4092608, 0.41083312, 0.4087006, 0.4201656, 0.40914744])
scores_mae = np.array([0.42773813, 0.43006033, 0.42706785, 0.4340189, 0.42810324])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.412 ± 0.004 | MAE: 0.429 ± 0.003


#### 2.2.3 Horizon 336

In [ ]:
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=336, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 243184
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth successfully transferred!

suggested_lr 0.000756463327554629
end-to-end finetuning
number of patches: 42
number of model params 243184
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-

In [ ]:
# Evaluate the results
scores_mse = np.array([0.42927605, 0.4306838, 0.43621603, 0.43227917, 0.42596078])
scores_mae = np.array([0.43994227, 0.44522637, 0.4438282, 0.4408326, 0.43753186])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.431 ± 0.003 | MAE: 0.441 ± 0.003


#### 2.2.4 Horizon 720

In [ ]:
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=720, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 501616
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_zero_model1.pth successfully transferred!

suggested_lr 0.000756463327554629
end-to-end finetuning
number of patches: 42
number of model params 501616
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-

In [ ]:
# Evaluate the results
scores_mse = np.array([0.48419365, 0.47366804, 0.46223584, 0.4858835, 0.46243554])
scores_mae = np.array([0.49013886, 0.47395447, 0.46860856, 0.48783943, 0.4800908])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.474 ± 0.010 | MAE: 0.480 ± 0.008


## 3. Gaussian Noise Addition

Unlike zero masking which completely removes information, Gaussian noise masking adds stochastic perturbations to the original signal:
$x_\text{masked} = x_\text{original} + \epsilon$, where $\epsilon \sim N(0, \sigma^2)$.

We evaluate three noise levels σ ∈ {0.1, 0.3, 0.5} to determine optimal magnitude. All candidates are tested with 10 pre-training epochs across 5 seeds on horizon 96.

### 3.1 Horizon 96

#### 3.1.1 Sigma 0.1

In [ ]:
# Pre-training
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.1 --pretrained_model_id 1 --seed 2019
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.1 --pretrained_model_id 2 --seed 2020
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.1 --pretrained_model_id 3 --seed 2021
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.1 --pretrained_model_id 4 --seed 2022
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.1 --pretrained_model_id 5 --seed 2023

args: Namespace(dset_pretrain='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, mask_ratio=0.4, use_gaussian_noise=1, noise_std=0.1, n_epochs_pretrain=10, lr=0.0001, pretrained_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 17260
suggested_lr 0.0027825594022071257
number of patches: 42
number of model params 17260
          epoch     train_loss     valid_loss           time
Better model found at epoch 0 with valid_loss value: 0.27906583994894524.
              0       0.914702       0.279066          00:01
Better model found at epoch 1 with valid_loss value: 0.09173365923403838.
              1       0.330932       0.091734          00:01
Better model found at epoch 2 with valid_loss value: 0.07121114679470096.
              2       0.223825       0.071211    

In [ ]:
# Fine-tuning
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth successfully transferred!

suggested_lr 0.0013219411484660286
end-to-end finetuning
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12

In [ ]:
scores_mse = np.array([0.36418888, 0.36402777, 0.3653059, 0.36440787, 0.36480618])
scores_mae = np.array([0.39490336, 0.39350045, 0.393983, 0.39243892, 0.39356822])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.365 ± 0.000 | MAE: 0.394 ± 0.001


#### 3.1.2 Sigma 0.3

In [ ]:
# Pre-training
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.3 --pretrained_model_id 1 --seed 2019
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.3 --pretrained_model_id 2 --seed 2020
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.3 --pretrained_model_id 3 --seed 2021
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.3 --pretrained_model_id 4 --seed 2022
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.3 --pretrained_model_id 5 --seed 2023

args: Namespace(dset_pretrain='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, mask_ratio=0.4, use_gaussian_noise=1, noise_std=0.3, n_epochs_pretrain=10, lr=0.0001, pretrained_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 17260
suggested_lr 0.0033516026509388406
number of patches: 42
number of model params 17260
          epoch     train_loss     valid_loss           time
Better model found at epoch 0 with valid_loss value: 0.26540308186995065.
              0       0.886092       0.265403          00:01
Better model found at epoch 1 with valid_loss value: 0.11123164123863892.
              1       0.330580       0.111232          00:01
Better model found at epoch 2 with valid_loss value: 0.09948137649727895.
              2       0.238355       0.099481    

In [ ]:
# Fine-tuning
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.3_model1.pth successfully transferred!

suggested_lr 0.0015922827933410938
end-to-end finetuning
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12

In [ ]:
scores_mse = np.array([0.3649411, 0.3671502, 0.3671684, 0.36698034, 0.36546987])
scores_mae = np.array([0.39679146, 0.3968761, 0.3956114, 0.3960654, 0.3940222])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.366 ± 0.001 | MAE: 0.396 ± 0.001


#### 3.1.3 Sigma 0.5

In [ ]:
# Pre-training
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.5 --pretrained_model_id 1 --seed 2019
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.5 --pretrained_model_id 2 --seed 2020
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.5 --pretrained_model_id 3 --seed 2021
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.5 --pretrained_model_id 4 --seed 2022
!python patchtst_pretrain.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --use_gaussian_noise 1 --noise_std 0.5 --pretrained_model_id 5 --seed 2023

args: Namespace(dset_pretrain='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, mask_ratio=0.4, use_gaussian_noise=1, noise_std=0.5, n_epochs_pretrain=10, lr=0.0001, pretrained_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 17260
suggested_lr 0.0033516026509388406
number of patches: 42
number of model params 17260
          epoch     train_loss     valid_loss           time
Better model found at epoch 0 with valid_loss value: 0.28975882675746184.
              0       0.917578       0.289759          00:01
Better model found at epoch 1 with valid_loss value: 0.1457408199618492.
              1       0.363757       0.145741          00:01
Better model found at epoch 2 with valid_loss value: 0.13726442840411804.
              2       0.273843       0.137264     

In [ ]:
# Fine-tuning
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=96, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.5_model1.pth successfully transferred!

suggested_lr 0.002310129700083158
end-to-end finetuning
number of patches: 42
number of model params 81664
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_

In [ ]:
scores_mse = np.array([0.36947724, 0.3653559, 0.36638978, 0.3681469, 0.36749917])
scores_mae = np.array([0.400238, 0.3955919, 0.39540088, 0.396572, 0.39631182])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.367 ± 0.001 | MAE: 0.397 ± 0.002


**Note:** σ = 0.1 achieves the best performance with the lowest mean MSE/MAE and smallest variance. Therefore, we will use this value for all remaining forecast horizons.

### 3.2 Horizon 192

In [ ]:
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --target_points 192 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=192, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 146272
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth successfully transferred!

suggested_lr 0.0006280291441834253
end-to-end finetuning
number of patches: 42
number of model params 146272
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_strid

In [ ]:
scores_mse = np.array([0.39742732, 0.3992837, 0.3976863, 0.39807117, 0.40277487])
scores_mae = np.array([0.41398013, 0.41467768, 0.4129167, 0.4129643, 0.41796625])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.399 ± 0.002 | MAE: 0.415 ± 0.002


### 3.3 Horizon 336

In [ ]:
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --target_points 336 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=336, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 243184
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth successfully transferred!

suggested_lr 0.000298364724028334
end-to-end finetuning
number of patches: 42
number of model params 243184
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride

In [ ]:
scores_mse = np.array([0.42710212, 0.42134792, 0.42921734, 0.42200008, 0.42789534])
scores_mae = np.array([0.43600264, 0.42992646, 0.43555441, 0.42949462, 0.43634143])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.426 ± 0.003 | MAE: 0.433 ± 0.003


### 3.4 Horizon 720

In [ ]:
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth --is_finetune 1 --finetuned_model_id 1 --seed 2019
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model2.pth --is_finetune 1 --finetuned_model_id 2 --seed 2020
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model3.pth --is_finetune 1 --finetuned_model_id 3 --seed 2021
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model4.pth --is_finetune 1 --finetuned_model_id 4 --seed 2022
!python patchtst_finetune.py --dset etth1 --target_points 720 --n_heads 4 --d_model 16 --d_ff 128 --pretrained_model ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model5.pth --is_finetune 1 --finetuned_model_id 5 --seed 2023

args: Namespace(is_finetune=1, is_linear_probe=0, dset_finetune='etth1', context_points=512, target_points=720, batch_size=64, num_workers=0, scaler='standard', features='M', percent=100, patch_len=12, stride=12, revin=1, affine=0, n_layers=3, n_heads=4, d_model=16, d_ff=128, dropout=0.2, head_dropout=0.2, n_epochs_finetune=20, lr=0.0001, pretrained_model='./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth', finetuned_model_id=1, model_type='based_model', seed=2019)
number of patches: 42
number of model params 501616
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride12_epochs-pretrain10_mask0.4_gnoise0.1_model1.pth successfully transferred!

suggested_lr 0.002310129700083158
end-to-end finetuning
number of patches: 42
number of model params 501616
weights from ./saved_models/etth1/masked_patchtst/based_model/patchtst_pretrained_cw512_patch12_stride

In [ ]:
scores_mse = np.array([0.4385243, 0.44501883, 0.4399612, 0.44110096, 0.44935784])
scores_mae = np.array([0.45992193, 0.46220887, 0.45895377, 0.45951146, 0.46723983])

print(f'MSE: {np.mean(scores_mse):.3f} ± {np.std(scores_mse):.3f} | MAE: {np.mean(scores_mae):.3f} ± {np.std(scores_mae):.3f}')

MSE: 0.443 ± 0.004 | MAE: 0.462 ± 0.003


## 4. Results

In [ ]:
print("Results:")
print(f"Avg. MSE improvement: {np.mean(((0.375-0.365)/0.375, (0.412-0.399)/0.412, (0.431-0.426)/0.431, (0.474-0.443)/0.474))*100:.3g}%")
print(f"Avg. MAE improvement: {np.mean(((0.405-0.394)/0.405, (0.429-0.415)/0.429, (0.441-0.433)/0.441, (0.480-0.462)/0.480))*100:.3g}%")

Results:
Avg. MSE improvement: 3.38%
Avg. MAE improvement: 2.89%


**Summary:**

| Horizon | Method | MSE | MAE |
|:-------:|:------:|:---:|:---:|
| **96** | Zero | 0.375 ± 0.002 | 0.405 ± 0.002 |
| | Gaussian | **0.365 ± 0.000** | **0.394 ± 0.001** |
| **192** | Zero | 0.412 ± 0.004 | 0.429 ± 0.003 |
| | Gaussian | **0.399 ± 0.002** | **0.415 ± 0.002** |
| **336** | Zero | 0.431 ± 0.003 | 0.441 ± 0.003 |
| | Gaussian | **0.426 ± 0.003** | **0.433 ± 0.003** |
| **720** | Zero | 0.474 ± 0.010 | 0.480 ± 0.008 |
| | Gaussian | **0.443 ± 0.004** | **0.462 ± 0.003** |

Gaussian noise masking consistently outperforms zero masking across all forecast horizons with average improvements of **3.38%** in MSE and **2.89%** in MAE on ETTh1. To have a complete validation of this approach, we'll evaluate both methods on two larger datasets: Electricity and Weather.

**Note:** Since all experiments use 100 pre-training epochs in the paper, ETTh1 experiments will be rerun with 100 pre-training epochs for consistency.